File: internal/domain/compose.go
package domain

type Compose struct {
	Version  string                   `yaml:"version"`
	Services map[string]ServiceConfig `yaml:"services"`
	Volumes  map[string]VolumeConfig  `yaml:"volumes"`
	Networks map[string]interface{}   `yaml:"networks"`
}

type ServiceConfig struct {
	Image           string            `yaml:"image,omitempty"`
	Build           string            `yaml:"build,omitempty"`
	Ports           []string          `yaml:"ports,omitempty"`
	Expose          []string          `yaml:"expose,omitempty"`
	Environment     []string          `yaml:"environment,omitempty"`
	EnvFile         []string          `yaml:"env_file,omitempty"`
	Volumes         []string          `yaml:"volumes,omitempty"`
	Networks        []string          `yaml:"networks,omitempty"`
	DependsOn       []string          `yaml:"depends_on,omitempty"`
	Links           []string          `yaml:"links,omitempty"`
	Restart         string            `yaml:"restart,omitempty"`
	ContainerName   string            `yaml:"container_name,omitempty"`
	Command         string            `yaml:"command,omitempty"`
	EntryPoint      string            `yaml:"entrypoint,omitempty"`
	Logging         string            `yaml:"logging,omitempty"`
	StdinOpen       bool              `yaml:"stdin_open,omitempty"`
	Tty             bool              `yaml:"tty,omitempty"`
	Labels          map[string]string `yaml:"labels,omitempty"`
	CapAdd          []string          `yaml:"cap_add,omitempty"`
	CapDrop         []string          `yaml:"cap_drop,omitempty"`
	DNS             []string          `yaml:"dns,omitempty"`
	DNSSearch       []string          `yaml:"dns_search,omitempty"`
	ExtraHosts      []string          `yaml:"extra_hosts,omitempty"`
	Hostname        string            `yaml:"hostname,omitempty"`
	Domainname      string            `yaml:"domainname,omitempty"`
	IPC             string            `yaml:"ipc,omitempty"`
	Privileged      bool              `yaml:"privileged,omitempty"`
	ReadOnly        bool              `yaml:"read_only,omitempty"`
	ShmSize         string            `yaml:"shm_size,omitempty"`
	SecurityOpt     []string          `yaml:"security_opt,omitempty"`
	Tmpfs           []string          `yaml:"tmpfs,omitempty"`
	Ulimits         []string          `yaml:"ulimits,omitempty"`
	User            string            `yaml:"user,omitempty"`
	WorkingDir      string            `yaml:"working_dir,omitempty"`
	Healthcheck     string            `yaml:"healthcheck,omitempty"`
	CgroupParent    string            `yaml:"cgroup_parent,omitempty"`
	Isolation       string            `yaml:"isolation,omitempty"`
	StopGracePeriod string            `yaml:"stop_grace_period,omitempty"`
	Sysctls         []string          `yaml:"sysctls,omitempty"`
	Secrets         []string          `yaml:"secrets,omitempty"`
	Deploy          string            `yaml:"deploy,omitempty"`
}

type VolumeConfig struct {
	Driver     string            `yaml:"driver,omitempty"`
	DriverOpts map[string]string `yaml:"driver_opts,omitempty"`
	External   bool              `yaml:"external,omitempty"`
	Name       string            `yaml:"name,omitempty"`
	Labels     map[string]string `yaml:"labels,omitempty"`
	Args       []string          `yaml:"args,omitempty"`
}

type NetworkConfig struct {
	Driver           string            `yaml:"driver,omitempty"`
	DriverOpts       map[string]string `yaml:"driver_opts,omitempty"`
	EnableIPv6       bool              `yaml:"enable_ipv6,omitempty"`
	IPAM             IPAMConfig        `yaml:"ipam,omitempty"`
	Internal         bool              `yaml:"internal,omitempty"`
	External         bool              `yaml:"external,omitempty"`
	Name             string            `yaml:"name,omitempty"`
	Labels           map[string]string `yaml:"labels,omitempty"`
	Attachable       bool              `yaml:"attachable,omitempty"`
	Ingress          bool              `yaml:"ingress,omitempty"`
	IPv6             bool              `yaml:"ipv6,omitempty"`
	IPv6Address      string            `yaml:"ipv6_address,omitempty"`
	IPv6PrefixLength int               `yaml:"ipv6_prefix_length,omitempty"`
}

type IPAMConfig struct {
	Driver string       `yaml:"driver,omitempty"`
	Config []IPAMSubnet `yaml:"config,omitempty"`
}

type IPAMSubnet struct {
	Subnet  string `yaml:"subnet,omitempty"`
	Gateway string `yaml:"gateway,omitempty"`
}

File: frontend/src/interfaces/compose.ts
export interface ICompose {
  name: string;
  description?: string;
  image?: string;
  createAt: string;
  updateAt: string;
}

export interface IComposeDetails extends ICompose {
  Version: string;
  Services: { [key: string]: ServiceConfig };
  Volumes: { [key: string]: VolumeConfig };
  Networks: { [key: string]: NetworkConfig };
}

export interface ServiceConfig {
  Image: string;
  Build: string;
  Ports: string[];
  Expose: string[];
  Environment: string[];
  EnvFile: string[];
  Volumes: string[];
  Networks: string[];
  DependsOn: string[];
  Links: string[];
  Restart: string;
  ContainerName: string;
  Command: string;
  EntryPoint: string;
  Logging: string;
  StdinOpen: boolean;
  Tty: boolean;
  Labels: { [key: string]: string };
  CapAdd: string[];
  CapDrop: string[];
  DNS: string[];
  DNSSearch: string[];
  ExtraHosts: string[];
  Hostname: string;
  Domainname: string;
  IPC: string;
  Privileged: boolean;
  ReadOnly: boolean;
  ShmSize: string;
  SecurityOpt: string[];
  Tmpfs: string[];
  Ulimits: string[];
  User: string;
  WorkingDir: string;
  Healthcheck: string;
  CgroupParent: string;
  Isolation: string;
  StopGracePeriod: string;
  Sysctls: string[];
  Secrets: string[];
  Deploy: string;
}

interface VolumeConfig {
  Driver: string;
  DriverOpts: { [key: string]: string };
  External: boolean;
  Name: string;
  Labels: { [key: string]: string };
  Args: string[];
}

interface NetworkConfig {
  Driver: string;
  DriverOpts: { [key: string]: string };
  EnableIPv6: boolean;
  IPAM: IPAMConfig;
  Internal: boolean;
  External: boolean;
  Name: string;
  Labels: { [key: string]: string };
  Attachable: boolean;
  Ingress: boolean;
  IPv6: boolean;
  IPv6Address: string;
  IPv6PrefixLength: number;
}

interface IPAMConfig {
  Driver: string;
  Config: IPAMSubnet[];
}

interface IPAMSubnet {
  Subnet: string;
  Gateway: string;
}
